utilities/transactions/pessimistic_transaction_db.h

```diff
@@ -161,6 +168,89 @@ class WritePreparedTxnDB : public PessimisticTransactionDB {
   Transaction* BeginTransaction(const WriteOptions& write_options,
                                 const TransactionOptions& txn_options,
                                 Transaction* old_txn) override;
+
+  // Check whether the transaction that wrote the value with seqeunce number seq
+  // is visible to the snapshot with sequence number snapshot_seq
+  bool IsInSnapshot(uint64_t seq, uint64_t snapshot_seq);
+  // Add the trasnaction with prepare sequence seq to the prepared list
+  void AddPrepared(uint64_t seq);
+  // Add the transaction with prepare sequence prepare_seq and commit sequence
+  // commit_seq to the commit map
+  void AddCommitted(uint64_t prepare_seq, uint64_t commit_seq);
+
+ private:
+  // A heap with the amortized O(1) complexity for erase. It uses one extra heap
+  // to keep track of erased entries that are not yet on top of the main heap.
+  class PreparedHeap {
+    std::priority_queue<uint64_t> prepared_txns_;
+    std::priority_queue<uint64_t> erased_prepared_txns_;
+
+   public:
+    bool empty() { return prepared_txns_.empty(); }
+    uint64_t top() { return prepared_txns_.top(); }
+    void push(uint64_t v) { prepared_txns_.push(v); }
+    void pop() {
+      prepared_txns_.pop();
+      while (!prepared_txns_.empty() && !erased_prepared_txns_.empty() &&
+             prepared_txns_.top() == erased_prepared_txns_.top()) {
```

Comment:
I see. Makes sense. In other words, what if we erase an entry that is already popped? I make the ::erase method ignore such entry. Thanks.